drivers/include/at.h

```diff
@@ -0,0 +1,202 @@
+} at_dev_t;
+
+
+/**
+ * @brief   Initialize AT device struct
+ *
+ * @param[in]   dev         struct to initialize
+ * @param[in]   uart        UART the device is connected to
+ * @param[in]   baudrate    baudrate of the device
+ * @param[in]   buf         input buffer
+ * @param[in]   bufsize     size of @p buf
+ *
+ * @returns     0 on success
+ * @returns     <0 otherwise
+ */
+int at_dev_init(at_dev_t *dev, uart_t uart, uint32_t baudrate, char *buf, size_t bufsize);
+
+/**
+ * @brief   simple command helper
+ *
+ * This function sends an AT command to the device and waits for "OK".
+ *
+ * @param[in]   dev     device to operate on
+ * @param[in]   command command string to send
+ * @param[in]   timeout timeout (in usec)
+ *
+ * @returns     0 when device answers "OK"
+ * @returns     <0 otherwise
+ */
+int at_send_cmd_wait_ok(at_dev_t *dev, const char *command, uint32_t timeout);
+
+/**
+ * @brief   send AT command, wait for a prompt
+ *
+ * This function will send the supplied @p command, then wait for the prompt (>)
+ * character and return
+ *
+ * @param[in]   dev     device to operate on
+ * @param[in]   command command string to send
+ * @param[in]   timeout timeout (in usec)
+ *
+ * @return      0 when prompt is received
+ * @return      <0 otherwise
+ */
+int at_send_cmd_wait_prompt(at_dev_t *dev, const char *command, uint32_t timeout);
+
+/**
+ * @brief   send AT command, wait for response
+ *
+ * This function will send the supplied @p command, then wait and return one
+ * line of response.
+ *
+ * A possible empty line will be skipped.
+ *
+ * @param[in]   dev         device to operate on
+ * @param[in]   command     command to send
+ * @param[out]  resp_buf    buffer for storing response
+ * @param[in]   len         len of @p buffer
+ * @param[in]   timeout     timeout (in usec)
+ *
+ * @returns     lenght of response on success
+ * @returns     <0 on error
+ */
+ssize_t at_send_cmd_get_resp(at_dev_t *dev, const char *command, char *resp_buf, size_t len, uint32_t timeout);
+
+/**
+ * @brief   send AT command, wait for multiline response
+ *
+ * This function will send the supplied @p command, then return all response
+ * lines until the device sends "OK".
+ *
+ * If a line starts with "ERROR" or "+CME ERROR:", or the buffer is full, the
+ * function returns -1.
+ *
+ * @param[in]   dev         device to operate on
+ * @param[in]   command     command to send
+ * @param[out]  resp_buf    buffer for storing response
+ * @param[in]   len         len of @p buffer
+ * @param[in]   timeout     timeout (in usec)
+ *
+ * @returns     lenght of response on success
+ * @returns     <0 on error
+ */
+ssize_t at_send_cmd_get_lines(at_dev_t *dev, const char *command, char *resp_buf, size_t len, uint32_t timeout);
+
+/**
+ * @brief   Expect bytes from device
+ *
+ * @param[in]   dev     device to operate on
+ * @param[in]   bytes   buffer containing bytes to expect
+ * @param[in]   len     number of bytes to expect
```

Comment:
Why do we need 'len'? The length of 'bytes' should be leading, otherwise you need a sanity check in the function.

Reply:
I don't quite get the question. ```len``` specifies the nr of bytes in ```bytes```. Maybe the docs are misleading?

Reply:
Why not rely on strlen(bytes) instead of having to pass a string and a length? In the implementation there is a while loop for 'len', but there is no sanity check whether there are that many chars in 'bytes'.

Without 'len' the usage could become as simple as:

    if (at_expect_bytes(dev, AT_END_OF_LINE "\r\n", timeout)) {

Bonus: in this case getting rid of the confusing sizeof()+1

Reply:
> Why not rely on strlen(bytes) instead of having to pass a string and a length?

the function is mostly used by other functions (quick grep counts five times). relying on strlen would lead to counting strings multiple times, for strings that are often static. having ```len``` as parameter allows the API user to decide whether to use strlen or not...



> there is no sanity check whether there are that many chars in 'bytes'

Reply:
I'm not impressed by that argument.
You can change the while condition in at_expect_bytes() to

    while (*bytes != 0) {

and then you won't need to do a strlen()

Reply:
That actually seems like a good idea. done.